### PR TITLE
fix erroneous failures in docker_compose due to deprecation warnings …

### LINKED ISSUE #60961: docker_compose treats Docker deprecation notice as a failure
##### SUMMARY
The latest Docker release has a bug that causes it to spit out a spurious deprecation warning when you pull an image. https://github.com/moby/moby/issues/39701

Ansible's `docker_compose` module treats that warning as a fatal error even though the operation actually succeeds.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
docker_compose

##### ANSIBLE VERSION

```paste below
ansible 2.8.4
  config file = /Users/koreth/work/terraform-configs/ansible/ansible.cfg
  configured module search path = ['/Users/koreth/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.8.4/libexec/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.4 (default, Jul  9 2019, 18:13:23) [Clang 10.0.1 (clang-1001.0.46.4)]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = True
ANSIBLE_SSH_CONTROL_PATH(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = %(directory)s/%%C
CACHE_PLUGIN(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = jsonfile
CACHE_PLUGIN_CONNECTION(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = /tmp/ansible-facts.json
CACHE_PLUGIN_TIMEOUT(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = 86400
DEFAULT_FORKS(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = 100
DEFAULT_GATHERING(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = smart
DEFAULT_ROLES_PATH(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = ['/Users/koreth/work/terraform-configs/ansible/thirdparty', '/Users/koreth/work/terraform-configs/ansible/roles', '/Users/koreth/work/terraform-configs/roles']
DEFAULT_VAULT_PASSWORD_FILE(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = /Users/koreth/work/terraform-configs/ansible/.vault-password
DEPRECATION_WARNINGS(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = False
DISPLAY_SKIPPED_HOSTS(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = False
RETRY_FILES_ENABLED(/Users/koreth/work/terraform-configs/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

Target OS is Ubuntu 18.04.1

Docker version 19.03.1, build 74b1e89

##### STEPS TO REPRODUCE

```yaml
- hosts: vagrant-test
  tasks:
    - docker_compose:
        project_name: test
        pull: yes
        definition:
          registrator:
            image: gliderlabs/registrator:v7
```

##### EXPECTED RESULTS

Docker image is pulled, service is launched, and task succeeds.

##### ACTUAL RESULTS

Docker image is pulled, service is launched, and task fails because of the unexpected warning message from Docker.

```paste below
TASK [docker_compose] *******************************************************************************
fatal: [vagrant-test]: FAILED! => {"changed": false, "module_stderr": "", "module_stdout": "v7: Pulling from gliderlabs/registrator\n[DEPRECATION NOTICE] registry v2 schema1 support will be removed in an upcoming release. Please contact admins of the docker.io registry NOW to avoid future disruption.\nDigest: sha256:0caaf84db8d645993964a0225c0485e7853135bf368393ce5a3a2c854379d476\nStatus: Image is up to date for gliderlabs/registrator:v7\n\n{\"services\": {\"registrator\": {\"test_registrator_1\": {\"cmd\": [], \"state\": {\"status\": \"running\", \"running\": true}, \"labels\": {\"com.docker.compose.service\": \"registrator\", \"com.docker.compose.config-hash\": \"5d06bcd558d0ce87526005a2d05a4c2b652d3952bdaeef9681d8552426863824\", \"com.docker.compose.project\": \"test\", \"com.docker.compose.version\": \"1.24.1\", \"com.docker.compose.oneoff\": \"False\", \"com.docker.compose.container-number\": \"1\"}, \"networks\": {\"bridge\": {\"macAddress\": \"02:42:ac:11:00:02\", \"globalIPv6PrefixLen\": 0, \"links\": null, \"IPPrefixLen\": 16, \"globalIPv6\": \"\", \"IPAddress\": \"172.17.0.2\", \"aliases\": []}}, \"image\": \"gliderlabs/registrator:v7\"}}}, \"invocation\": {\"module_args\": {\"project_src\": null, \"hostname_check\": false, \"recreate\": \"smart\", \"docker_host\": \"unix://var/run/docker.sock\", \"remove_images\": null, \"client_key\": null, \"scale\": null, \"nocache\": false, \"remove_volumes\": false, \"state\": \"present\", \"build\": false, \"client_cert\": null, \"api_version\": \"auto\", \"files\": null, \"project_name\": \"test\", \"ca_cert\": null, \"pull\": true, \"tls_hostname\": \"localhost\", \"dependencies\": true, \"services\": null, \"tls\": false, \"definition\": {\"registrator\": {\"image\": \"gliderlabs/registrator:v7\"}}, \"remove_orphans\": false, \"restarted\": false, \"ssl_version\": null, \"stopped\": false, \"timeout\": 10, \"debug\": false, \"validate_certs\": false}}, \"changed\": true, \"ansible_facts\": {\"registrator\": {\"test_registrator_1\": {\"cmd\": [], \"state\": {\"status\": \"running\", \"running\": true}, \"labels\": {\"com.docker.compose.service\": \"registrator\", \"com.docker.compose.config-hash\": \"5d06bcd558d0ce87526005a2d05a4c2b652d3952bdaeef9681d8552426863824\", \"com.docker.compose.project\": \"test\", \"com.docker.compose.version\": \"1.24.1\", \"com.docker.compose.oneoff\": \"False\", \"com.docker.compose.container-number\": \"1\"}, \"networks\": {\"bridge\": {\"macAddress\": \"02:42:ac:11:00:02\", \"globalIPv6PrefixLen\": 0, \"links\": null, \"IPPrefixLen\": 16, \"globalIPv6\": \"\", \"IPAddress\": \"172.17.0.2\", \"aliases\": []}}, \"image\": \"gliderlabs/registrator:v7\"}}}}\n", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 0}
```

### PR DESCRIPTION
**Pull request by sluther**
…from docker (#60961)

##### SUMMARY

Consistently capture stdout and stderr to prevent any output from causing erroneous failures

fixes #60961
##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

docker_compose

### CODE REVIEW

**Review by felixfontein — APPROVED** [review]
~~LGTM~~ wait, one detail is missing

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 1]
```diff
@@ -866,9 +866,12 @@ def cmd_pull(self):
...
 
+                out_redir_name, err_redir_name = make_redirection_tempfiles()
                 # pull the image
                 try:
-                    service.pull(ignore_pull_failures=False)
+                    with stdout_redirector(out_redir_name):
+                        with stderr_redirector(err_redir_name):
+                            service.pull(ignore_pull_failures=False)
                 except Exception as exc:
                     self.client.fail("Error: pull failed with %s" % str(exc))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]: The other places where `make_redirection_tempfiles()` is used are doing the following:
```suggestion
                    fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
                                                   msg_format="Error: pull failed with %s")
                    self.client.fail(**fail_reason)
                else:
                    cleanup_redirection_tempfiles(out_redir_name, err_redir_name)
```
That also cleans up the output files in both error and non-error case.

(Same for the other places.)

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/docker/docker_compose.py (hunk 1 above)
Comment (sluther) [reply lib/ansible/modules/cloud/docker/docker_compose.py]: Ah! I was actually wondering about this. I'll get that added as well!

**Comment by sluther** [comment]
Should this also get added to line https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_compose.py#L934?

**Comment by sluther** [comment]
Also, not sure how to handle the change for this line: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_compose.py#L1063

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 2]
```diff
@@ -913,9 +920,12 @@ def cmd_build(self):
...
 
+                    out_redir_name, err_redir_name = make_redirection_tempfiles()
                     # build the image
                     try:
-                        new_image_id = service.build(pull=self.pull, no_cache=self.nocache)
+                        with stdout_redirector(out_redir_name):
+                            with stderr_redirector(err_redir_name):
+                                new_image_id = service.build(pull=self.pull, no_cache=self.nocache)
                     except Exception as exc:
                         self.client.fail("Error: build failed with %s" % str(exc))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                        fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
                                                       msg_format="Error: build failed with %s")
                        self.client.fail(**fail_reason)
                    else:
                        cleanup_redirection_tempfiles(out_redir_name, err_redir_name)
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 3]
```diff
@@ -1037,8 +1050,11 @@ def cmd_scale(self):
...
                     service_res['scale'] = scale - len(containers)
                     if not self.check_mode:
+                        out_redir_name, err_redir_name = make_redirection_tempfiles()
                         try:
-                            service.scale(scale)
+                            with stdout_redirector(out_redir_name):
+                                with stderr_redirector(err_redir_name):
+                                    service.scale(scale)
                         except Exception as exc:
                             self.client.fail("Error scaling %s - %s" % (service.name, str(exc)))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                            fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
                                                           msg_format="Error scaling {0} - %s".format(service.name))
                            self.client.fail(**fail_reason)
                        else:
                            cleanup_redirection_tempfiles(out_redir_name, err_redir_name)
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 4]
```diff
@@ -946,8 +956,11 @@ def cmd_down(self):
...
         if not self.check_mode and result['changed']:
             image_type = image_type_from_opt('--rmi', self.remove_images)
+            out_redir_name, err_redir_name = make_redirection_tempfiles()
             try:
-                self.project.down(image_type, self.remove_volumes, self.remove_orphans)
+                with stdout_redirector(out_redir_name):
+                    with stderr_redirector(err_redir_name):
+                        self.project.down(image_type, self.remove_volumes, self.remove_orphans)
             except Exception as exc:
                 self.client.fail("Error stopping project - %s" % str(exc))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
                                               msg_format="Error stopping project - %s")
                self.client.fail(**fail_reason)
            else:
                cleanup_redirection_tempfiles(out_redir_name, err_redir_name)
```

**Review by felixfontein — COMMENTED** [review]
About https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_compose.py#L934: I don't think this needs any change, since stdout/stderr isn't captured in there. Or do you mean whether stdout/stderr should also be captured for `service.image()` calls?

File: lib/ansible/modules/cloud/docker/docker_compose.py (hunk 3 above)
Comment (sluther) [reply lib/ansible/modules/cloud/docker/docker_compose.py]: Not sure why github isn't saying that this is outdated... but this is fixed.

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 5]
```diff
@@ -946,10 +960,17 @@ def cmd_down(self):
...
+            out_redir_name, err_redir_name = make_redirection_tempfiles()
             try:
-                self.project.down(image_type, self.remove_volumes, self.remove_orphans)
+                with stdout_redirector(out_redir_name):
+                    with stderr_redirector(err_redir_name):
+                        self.project.down(image_type, self.remove_volumes, self.remove_orphans)
             except Exception as exc:
-                self.client.fail("Error stopping project - %s" % str(exc))
+                fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
+                                                msg_format="Error stopping project - %s)
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                                                msg_format="Error stopping project - %s")
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 6]
```diff
@@ -913,11 +920,18 @@ def cmd_build(self):
...
                     # build the image
                     try:
-                        new_image_id = service.build(pull=self.pull, no_cache=self.nocache)
+                        with stdout_redirector(out_redir_name):
+                            with stderr_redirector(err_redir_name):
+                                new_image_id = service.build(pull=self.pull, no_cache=self.nocache)
                     except Exception as exc:
-                        self.client.fail("Error: build failed with %s" % str(exc))
+                        fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
+                                                        msg_format="Error: build failed with %s")
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                                                       msg_format="Error: build failed with %s")
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 7]
```diff
@@ -946,10 +960,17 @@ def cmd_down(self):
...
+            out_redir_name, err_redir_name = make_redirection_tempfiles()
             try:
-                self.project.down(image_type, self.remove_volumes, self.remove_orphans)
+                with stdout_redirector(out_redir_name):
+                    with stderr_redirector(err_redir_name):
+                        self.project.down(image_type, self.remove_volumes, self.remove_orphans)
             except Exception as exc:
-                self.client.fail("Error stopping project - %s" % str(exc))
+                fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
+                                                msg_format="Error stopping project - %s")
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                                               msg_format="Error stopping project - %s")
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 8]
```diff
@@ -979,7 +1000,7 @@ def cmd_stop(self, service_names):
                         self.project.stop(service_names=service_names, timeout=self.timeout)
             except Exception as exc:
                 fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
-                                               msg_format="Error stopping project %s")
+                                                msg_format="Error stopping project %s")
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                                               msg_format="Error stopping project %s")
```

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 9]
```diff
@@ -1037,10 +1058,18 @@ def cmd_scale(self):
...
+                        out_redir_name, err_redir_name = make_redirection_tempfiles()
                         try:
-                            service.scale(scale)
+                            with stdout_redirector(out_redir_name):
+                                with stderr_redirector(err_redir_name):
+                                    service.scale(scale)
                         except Exception as exc:
                             self.client.fail("Error scaling %s - %s" % (service.name, str(exc)))
+                            fail_reason = get_failure_info(exc, out_redir_name, err_redir_name,
+                                                            msg_format="Error scaling {0} - %s".format(service.name))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]:
```suggestion
                                                           msg_format="Error scaling {0} - %s".format(service.name))
```

**Review by felixfontein — COMMENTED** [review]
The next line needs to be aligned with the first character after `(`.

File: lib/ansible/modules/cloud/docker/docker_compose.py [hunk 10]
```diff
@@ -1037,10 +1058,18 @@ def cmd_scale(self):
...
                     service_res['scale'] = scale - len(containers)
                     if not self.check_mode:
+                        out_redir_name, err_redir_name = make_redirection_tempfiles()
                         try:
-                            service.scale(scale)
+                            with stdout_redirector(out_redir_name):
+                                with stderr_redirector(err_redir_name):
+                                    service.scale(scale)
                         except Exception as exc:
                             self.client.fail("Error scaling %s - %s" % (service.name, str(exc)))
```
Comment (felixfontein) [inline lib/ansible/modules/cloud/docker/docker_compose.py]: This line has to go:
```suggestion
```

**Review by felixfontein — COMMENTED** [review]
Looks good besides that :)

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@sluther thanks for fixing this! Do you want to create a backport to stable-2.8 and stable-2.9, or should I do that?

**Comment by sluther** [comment]
I can make a backport for 2.8 and 2.9 this weekend.

**Pull request merged**